### PR TITLE
Constant time map stats

### LINKED ISSUE #4022: IMap stats showing negative values
I'm using hazelcast 2.6.7 and after heavy usage of the cache, the number of misses starts showing up negative. Specifically, it is being reported as the negative of the number of hits (e.g. if hits = 20, misses = -20). I'm getting the map stats as follows:

    LocalMapStats mapStats = HazelcastCache.instance.getMap("myMap").getLocalMapStats();
    long hits = mapStats.getHits();
    long misses = mapStats.getMisses();

### PR DESCRIPTION
**Pull request by muratayan**
`LocalMapStats` logic doesn't change. Complexity improved from per entry O(N) to per partition O(p) = O(1)

fixes #4022

### CODE REVIEW

**Comment by emrahkocaman** [comment]
👍

File: hazelcast/src/main/java/com/hazelcast/map/impl/recordstore/AbstractRecordStore.java [hunk 1]
```diff
@@ -208,4 +218,48 @@ public void disposeDeferredBlocks() {
...
+        setLastUpdateTime(now);
+
+        if (!newRecord) {
+            updateStatsOnGet(now);
+        }
+    }
+
+    protected void updateStatsOnGet(long now) {
+        setLastAccessTime(now);
+        increaseHits();
```
Comment (ahmetmircik) [inline hazelcast/src/main/java/com/hazelcast/map/impl/recordstore/AbstractRecordStore.java]: we also need to decrease hits in the event of entry removal.

Comment (emrahkocaman) [inline hazelcast/src/main/java/com/hazelcast/map/impl/recordstore/AbstractRecordStore.java]: I'm not sure about that. We're trying to reach the total hit count of a map, right? Imagine that a map is cleared, then this means the hit count is 0 now? This should not be the case, IMO.

**Comment by muratayan** [comment]
Updated

File: hazelcast/src/main/java/com/hazelcast/map/impl/recordstore/AbstractRecordStore.java (hunk 1 above)
Comment (enesakar) [inline hazelcast/src/main/java/com/hazelcast/map/impl/recordstore/AbstractRecordStore.java]: agree with @emrahkocaman

File: hazelcast/src/main/java/com/hazelcast/map/impl/recordstore/AbstractRecordStore.java [hunk 2]
```diff
@@ -208,4 +218,72 @@ public void disposeDeferredBlocks() {
...
+    public void setLastUpdateTime(long time) {
+        this.lastUpdate = Math.max(this.lastUpdate, time);
+    }
+
+
+    protected void updateStatsOnPut(boolean newRecord, long now) {
+        setLastUpdateTime(now);
+
+        if (!newRecord) {
+            updateStatsOnGet(now);
```
Comment (enesakar) [inline hazelcast/src/main/java/com/hazelcast/map/impl/recordstore/AbstractRecordStore.java]: Yes we can update the access time but I think we should not increase hits on update operation. wdyt? @ahmetmircik @emrahkocaman

File: hazelcast/src/main/java/com/hazelcast/map/impl/recordstore/AbstractRecordStore.java (hunk 1 above)
Comment (ahmetmircik) [inline hazelcast/src/main/java/com/hazelcast/map/impl/recordstore/AbstractRecordStore.java]: To me, if this PR is trying to preserve existing behaviour (it may be correct or not, discussion for  that is out of scope) and is only providing O(1) calculation for stats, `decrease` should be there because in current impl. after `map#clear`, there is no hit.

File: hazelcast/src/main/java/com/hazelcast/map/impl/recordstore/AbstractRecordStore.java (hunk 2 above)
Comment (ahmetmircik) [inline hazelcast/src/main/java/com/hazelcast/map/impl/recordstore/AbstractRecordStore.java]: makes sense

Comment (muratayan) [inline hazelcast/src/main/java/com/hazelcast/map/impl/recordstore/AbstractRecordStore.java]: Makes sense but won't do it because we are trying to preserve existing behaviour.

**Comment by ahmetmircik** [comment]
👍
